### ISSUE TITLE
Backbone-associations + node fails to update

### ISSUE THREAD

**Issue by joeybaker** [dhruvaray/backbone-associations#72]
I'm having a hell of time pinning this issue down, but I thought I'd share progress in the hopes that someone smarter than me can figure it out.

TL;DR: when using backbone-associations in a node.js app, updates to a sub-model with a direct `save` call will fail to populate the sub-model attributes until sometime after the `success` callback is fired.

I've got backbone.js running in a node.js app. `Backbone.sync` is [overridden](http://github.com/joeybaker/wheelhouse-couch), so a `save` call goes to a db. The following code produces a very weird result: the model's sub attributes are not updated until after the model's `success` callback.

``` js
// code both client and server can see
var Model = Backbone.AssociatedModel.extend({
    relations: [
      {
        type: Backbone.Many
        , key: 'subModels'
        , relatedModel: subModel
      }
    ]
    , defaults: {
      subModels: []
    }
  })
  , subModel = Backbone.AssociatedModel.extend({
    defaults: {
      value: true
    }
  })
  , model = new Model({subModels: [{}]})

// client side only
model.save({subModels: [{value: 'updated!'}]})

// server only code

// data === {subModels: [{value: 'updated!'}]}
model.save(data, {
  success: function(modelUpdated){
    console.log(model.toJSON()) // model is updated

    db.get(model.id, function(err, doc){
      console.log(doc) // model not updated
    })
  }))
  }
  , wait: true
})


```

However, a slight modification to the `save` call gets everything working happily

``` js
    model.set(data, {silent: true}).save(null)
```

I'm totally unclear why calling a `set` before a `save` is a fix since this what `save` is essentially doing anyway…

**Comment by jdkanani** [comment]
I am not sure If I understand you correctly, but here I have one
question before I start debugging.

Is there any particular reason to pass `wait: true` as option?

>   Backbone doc says:
>   Pass `wait: true` if you'd like to wait for the server before
>   setting the new attribute on the model.

Let's have a look what Backbone is doing internally.

with `wait`

```
sync(method, this, {
    success: function(response) {
        ...
         var serverAttrs = model.parse(response, options);
         this.set(serverAttrs);
         ...
     }
});
```

without `wait` (This is exactly what you are doing to make it work):
    model.set(data).save(null, options);

Let me know if it works for you or I will take a deep dive into this.

**Comment by dhruvaray** [comment]
Can we assume that the issue is fixed for you? If you think otherwise, please re-open

**Comment by joeybaker** [comment]
Sorry, somehow I never got emailed when you commented. Apologies for the late reply. My issue isn't solved.

The reason I'm passing `wait: true` is because I've got `Backbone.sync` setup to save to a database, and I'd like to ensure the data actually saves before moving on. …Irrespective of that need though, it strikes me as "broken" that backbone-associations won't work with the `wait` option passed.

I'd re-open, but I don't have that ability :\

**Comment by jdkanani** [comment]
[Here](https://c9.io/jdkanani/demo-project) is the live demo for your code (with some changes for `mongodb`). You can use those files and run this code:

```
$ node main.js
```

You can check behavior with and without `wait: true` flag. If you pass `wait` flag, Backbone doesn't set `data` to `model` before saving it to db. It will wait for `save` success.

In any case, If you fetch `model` from db immediately after calling `save` - you can get yourself into race condition. Therefore, in this scenario, you should fetch/check your model in the callback of `save`.

Once, you visit the site, I can add you as a read-write collaborator.

**Comment by joeybaker** [comment]
@jdkanani Sorry, I should have been more clear in my code. I've updated it to clarify. The race condition you specified is not the root cause.

**Comment by jdkanani** [comment]
Backbone sets "proposed" attributes in `this.attributes`  temporarily before calling `toJSON` in save in case of `wait`.

``` javascript
// Set temporary attributes if `{wait: true}`
if (attrs && options.wait) {
    this.attributes = _.extend({}, attributes, attrs); // direct set
}

// on success of server save

// Ensure attributes are restored during synchronous saves.
model.attributes = attributes;
var serverAttrs = model.parse(resp, options);
if (options.wait) serverAttrs = _.extend(attrs || {}, serverAttrs);
if (_.isObject(serverAttrs) && !model.set(serverAttrs, options)) {
    return false;
}
```

As it sets `this.attributes` directly we can't have control over `set`.

`sync` method with `{wait: true}`:  

``` js
sync:  function (method, model, options) {
    var data = model.toJSON(); // This fails as `subModels` is not an associated collection.

    // model

    //{
    //    cid: "c1",
    //    changed: {},
    //    _previousAttributes: {},
    //    _pending: false,
    //    attributes: {
    //        name: 'parent',
    //        subModels: [            // Simple array (not nested models)
    //            {  
    //                value: 'updated!'
    //            }
    //        ]
    //    }
    //}


    db.save(data, function(e, result){
        if (e) return e;
        options.success && options.success(result);
    });
}

```

I have added fix for this `toJSON` problem with [this commit](https://github.com/dhruvaray/backbone-associations/commit/64cd3e64a90bb6615d52dfd05e09659fcbeb7503). You can try out and check if you are getting intended `json` data from model while syncing with db.

Now, problem for consistency of `attributes` remains same for backbone also - see bug https://github.com/jashkenas/backbone/issues/2799

Let me know if you need any help to take down this problem.
